File: app/src/main/java/com/example/englishapp/di/AppBindingModule.java
package com.example.englishapp.di;

import com.example.englishapp.feature.scan.data.repository.ScanRepositoryImpl;
import com.example.englishapp.feature.scan.domain.repository.ScanRepository;

import dagger.Binds;
import dagger.Module;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

import javax.inject.Singleton;

@Module
@InstallIn(SingletonComponent.class)
public abstract class AppBindingModule {

    @Binds
    @Singleton
    public abstract ScanRepository bindScanRepository(ScanRepositoryImpl impl);
}

File: app/src/main/java/com/example/englishapp/di/AppBindingModule.kt
package com.project.englishapp.di

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStore
import com.example.englishapp.core.common.DefaultDispatcherProvider
import com.example.englishapp.core.common.DispatcherProvider
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import javax.inject.Qualifier
import javax.inject.Singleton

@Qualifier
@Retention(AnnotationRetention.BINARY)
annotation class IoDispatcher

@Qualifier
@Retention(AnnotationRetention.BINARY)
annotation class DefaultDispatcher

@Qualifier
@Retention(AnnotationRetention.BINARY)
annotation class MainDispatcher

private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "scan_learn_settings")

@Module
@InstallIn(SingletonComponent::class)
object AppBindingModule {

    @Provides
    @Singleton
    fun provideDispatcherProvider(): DispatcherProvider = DefaultDispatcherProvider()

    @DefaultDispatcher
    @Provides
    fun provideDefaultDispatcher(): CoroutineDispatcher = Dispatchers.Default

    @MainDispatcher
    @Provides
    fun provideMainDispatcher(): CoroutineDispatcher = Dispatchers.Main

    @Provides
    @Singleton
    fun provideDataStore(@ApplicationContext context: Context): DataStore<Preferences> {
        return context.dataStore
    }
}